File: shared/presets.ts
// DATA catalog of map presets. A preset biases worldgen: it nudges the sea level
// and the moisture banding so the same generator yields a recognizably different
// land — a wet "Verdant" map, a parched "Desert", a sea-broken "Archipelago", a
// rugged "Highlands". Pure data; terrain.ts reads a MapBias and shifts its
// thresholds. The module stores the chosen preset id alongside the seed so the
// client recomputes identical land.
//
// Biasing the SAME classify() (rather than forking generators) keeps one source
// of truth for biome bands and guarantees module/client agreement.

export interface MapBias {
  seaShift: number; // + raises the waterline (more sea), - lowers it (more land)
  moistShift: number; // + wetter (grass/forest), - drier (desert/dunes)
  elevGain: number; // multiplies land relief — higher = taller hills/mountains
}

export interface MapPreset {
  id: string;
  label: string;
  description: string;
  bias: MapBias;
}

export const NEUTRAL_BIAS: MapBias = { seaShift: 0, moistShift: 0, elevGain: 1 };

export const MAP_PRESETS: MapPreset[] = [
  {
    id: 'continental',
    label: 'Continental',
    description: 'Balanced land with a fair mix of every biome.',
    bias: NEUTRAL_BIAS,
  },
  {
    id: 'verdant',
    label: 'Verdant',
    description: 'Wet, fertile country — broad grassland and deep forest.',
    bias: { seaShift: -0.02, moistShift: 0.16, elevGain: 0.9 },
  },
  {
    id: 'desert',
    label: 'Arabian Desert',
    description: 'Parched dunes and sand, sparse oases by the water.',
    bias: { seaShift: 0.0, moistShift: -0.2, elevGain: 0.85 },
  },
  {
    id: 'highlands',
    label: 'Highlands',
    description: 'Rugged uplands — towering hills, mountains, and snow.',
    bias: { seaShift: -0.03, moistShift: 0.02, elevGain: 1.45 },
  },
  {
    id: 'archipelago',
    label: 'Archipelago',
    description: 'A sea of scattered islands — control the straits.',
    bias: { seaShift: 0.1, moistShift: 0.06, elevGain: 1.0 },
  },
];

export function mapPresetById(id: string): MapPreset {
  return MAP_PRESETS.find((p) => p.id === id) ?? MAP_PRESETS[0];
}

export function mapPresetByIndex(index: number): MapPreset {
  return MAP_PRESETS[((index % MAP_PRESETS.length) + MAP_PRESETS.length) % MAP_PRESETS.length];
}

export function biasOf(presetId: string): MapBias {
  return mapPresetById(presetId).bias;
}

File: saladin-bevy/crates/sim/src/presets.rs
use crate::math::Fx;

/// A map preset biases worldgen so the same generator yields recognizably
/// different land. `elev_gain` multiplies render relief (client only); the shift
/// fields nudge classification thresholds.
#[derive(Clone, Copy, Debug)]
pub struct MapBias {
    pub sea_shift: Fx,
    pub moist_shift: Fx,
    pub elev_gain: Fx,
}

#[derive(Clone, Copy, Debug)]
pub struct MapPreset {
    pub id: &'static str,
    pub label: &'static str,
    pub description: &'static str,
    pub bias: MapBias,
}

pub const NEUTRAL_BIAS: MapBias =
    MapBias { sea_shift: crate::fx!("0"), moist_shift: crate::fx!("0"), elev_gain: crate::fx!("1") };

pub const MAP_PRESETS: [MapPreset; 5] = [
    MapPreset {
        id: "continental",
        label: "Continental",
        description: "Balanced land with a fair mix of every biome.",
        bias: NEUTRAL_BIAS,
    },
    MapPreset {
        id: "verdant",
        label: "Verdant",
        description: "Wet, fertile country — broad grassland and deep forest.",
        bias: MapBias { sea_shift: crate::fx!("-0.02"), moist_shift: crate::fx!("0.16"), elev_gain: crate::fx!("0.9") },
    },
    MapPreset {
        id: "desert",
        label: "Arabian Desert",
        description: "Parched dunes and sand, sparse oases by the water.",
        bias: MapBias { sea_shift: crate::fx!("0"), moist_shift: crate::fx!("-0.2"), elev_gain: crate::fx!("0.85") },
    },
    MapPreset {
        id: "highlands",
        label: "Highlands",
        description: "Rugged uplands — towering hills, mountains, and snow.",
        bias: MapBias { sea_shift: crate::fx!("-0.03"), moist_shift: crate::fx!("0.02"), elev_gain: crate::fx!("1.45") },
    },
    MapPreset {
        id: "archipelago",
        label: "Archipelago",
        description: "A sea of scattered islands — control the straits.",
        bias: MapBias { sea_shift: crate::fx!("0.1"), moist_shift: crate::fx!("0.06"), elev_gain: crate::fx!("1.0") },
    },
];

pub fn map_preset_by_id(id: &str) -> &'static MapPreset {
    MAP_PRESETS.iter().find(|p| p.id == id).unwrap_or(&MAP_PRESETS[0])
}

pub fn map_preset_by_index(index: i32) -> &'static MapPreset {
    let n = MAP_PRESETS.len() as i32;
    let i = ((index % n) + n) % n;
    &MAP_PRESETS[i as usize]
}

pub fn bias_of(preset_id: &str) -> MapBias {
    map_preset_by_id(preset_id).bias
}
